Добавлены скрипты для отрисовки карточек товара и показа модальных окон

eimwe opened the pull request:
Домашнее задание №5 выполнено

eimwe committed:
Создал мок массивы с данными для карточек товара

1. Массивы с объектами поместил в файл shopcards.js
2. Импрортировал массивы в catalog.js

eimwe committed:
Добавил темплейт карточки товара в разметку Главной

1. Исправил разметку для контейнера карточек в файле index.html (output вместо div)
2. Обернул шаблон карточки товара в template c идентификатором shopcard-template
3. Изменил обертку для контента карточки с div на a, т.к. в output нельзя вкладывать блочные элементы
*4. Добавил внешние отступы по правому и левому краю для изображений

* - в карточках с нефиксированной шириной изображения по умолчанию не были отцентрированы

eimwe committed:
Добавил функцию для отрисовки карточек товаров

1. customizeCardTemplate() - создает карточки на основе шаблона, подставляя значения, переданные в итераторы
2. forEach() итерируется по массивам объектов apparel и accessories и наполняет карточки значениями из массива

eimwe committed:
Добавил исправления для корректного рендеринга карточек

1. В index.html удалил из разметки шаблона темплейта карточки статичные данные
2. В файле shopcards.js заменил html-сущности &quot; на обычные кавычки
3. В стилях main.scss увеличил у контейнера карточек внешний отступ снизу на 40px

eimwe committed:
Добавил картинки для карточек товара

* также исправил стоимости некоторых товаров в файле shopcards.js

eimwe committed:
Вывел новые товары в начало списка

1. Создал отдельную функцию для рендеринга карточек - renderCards()
2. Добавил в функцию renderCards() условие для вывода карточек с плашкой "New" в начало списка товаров
3. Передал функцию renderCards() в итераторы

eimwe committed:
Добавил обработчик событий на переключение кнопок категорий

1. Добавил радио-кнопкам атрибуты data-key в файле index.html
2. Добавил слушателей onchange радио-кнопкам, получил data-key по каждой кнопке
3. Объединил категории 'Одежда' и 'Аксессуары' в массив everything
4. Сделал переменную контейнера карточек (cardContainer) глобальной
5. Добавил условия для рендеринга карточек в зависимости от выбранной категории

eimwe committed:
Добавил функцию очистки контейнера от карточек при переключении между категориями

1. Условия для очистки описал в функции cleanCardContainer()
2. Передал функцию cleanCardContainer() в условия обработчика событий

eimwe committed:
Добавил функцию рендеринга карточек по умолчанию

1. Создал функцию renderCardsByCategory(), принимающую переменную category
2. Передал функцию renderCardsByCategory() в обработчик событий, передавая в функцию категорию товара в зависимости от описанных в обработчике условий
3. Вызвал функцию renderCardsByCategory() с параметром 'everything' для отрисовки всех карточек при первоначальной загрузке страницы
4. Переместил объявление переменной с массивом всех радио-кнопок в начало скрипта для единообразия
5. Переместил функцию очистки контейнера cleanCardContainer() перед проверкой условий, чтобы избежать дублирования кода

eimwe committed:
Добавил в массив данные для модального окна

1. Детали (details)
2. Полноразмерные фото для галереи (fullsized)
3. Превью для галереи (thumbs)

eimwe committed:
Добавил картинки для галереи модального окна

eimwe committed:
Добавил темплейт модального окна в разметку Главной

eimwe committed:
Добавил функцию customizeOrderModalTemplate()

Функция выполняет:
1. Клонирование модального окна из темплейта
2. Создание узлов для галереи
3. Добавление переменных из массива данных карточек товаров
4. Возврат полученного клона

eimwe committed:
Добавил функцию отрисовки модального окна renderOrderModals()

Функция выполняет:
1. Заполнение данных контента модального окна в зависимости от переданного индекса массива
2. Добавление узла модального окна на страницу перед хедером
3. Возврат узла модального окна

eimwe committed:
Добавил функцию показа модального окна

Для реализации функции потребовалось внести следующие изменения в скрипт:
1. Создать функцию showOrderModal()
2. Вернуть узлы карточек (renderedCards) из функции renderCardsByCategory()
3. Передать параметром функцию renderCardsByCategory() в showOrderModal()

eimwe committed:
Добавил функцию закрытия модального окна

eimwe committed:
Исправил стили для модального окна

1. Добавил позиционирование и ширину для подложки для центровки модального окна
2. Убрал глобальные стили из файла order.scss, подключил файл с переменными scss
3. Импортировал файл стилей order.scss в catalog.js

eimwe committed:
Добавил скрипт для переключения слайдов в галерее

eimwe committed:
Добавил мелкие исправления

1. Переименовал фото в директории src/assets/images/gallery/ для корректного показа в галерее товара id: 4
2. Переделал details в файле shopcards.js у товара id: 9, чтобы избежать ненужного переноса строки

kerim1505k commented on src/js/catalog.js:
```diff
@@ -1 +1,193 @@
+        renderCards(card);
+    });
+
+    const renderedCards = document.querySelectorAll('.shop-card');
+
+    return renderedCards;
+}
```
Можно сразу return document.querySelectorAll('.shop-card');

eimwe replied:
"+" Исправлено

kerim1505k commented on src/js/catalog.js:
```diff
@@ -1 +1,193 @@
+    } = card;
+    const customizedCard = customizeCardTemplate(id, img, title, price, isTagged);
+
+    cardContainer.appendChild(customizedCard);
```
Нужна проверка на cardContainer

Есть shorthand проверки

```
cardContainer?.appendChild(customizedCard);

```

eimwe replied:
"+" Исправлено

kerim1505k commented on src/js/catalog.js:
```diff
@@ -1 +1,193 @@
+const showOrderModal = (renderedCards) => {
+    renderedCards.forEach((card) => {
+        card.addEventListener('click', (event) => {
+            const clickedCard = event.target;
```
Тут же можно сразу напрямую использовать id из card.dataset

eimwe replied:
"+" Исправлено

kerim1505k commented on src/js/catalog.js:
```diff
@@ -1 +1,193 @@
+    }
+};
+
+const renderOrderModals = (card, id) => {
```
Тут необязательно перекидывать постоянно общий массив "everything" с товарами. Он у тебя объявлен глобально, можешь просто заюзать его здесь, никаких модификаций перед попаданием сюда он не проходит, поэтому нет смысла передавать его как параметр.

eimwe replied:
"+" Исправлено

kerim1505k commented on src/js/catalog.js:
```diff
@@ -1 +1,193 @@
+    const {
+        title, price, details, fullsized, thumbs,
+    } = card[id];
+    const customizedOrderModal = customizeOrderModalTemplate(title, price, details, fullsized, thumbs);
```
Нужны проверки на передаваемые данные. В реальных проектах чего-то может и не быть.

eimwe replied:
Похоже, что я не понимаю, как мне следует проверять данные. Я написал функцию проверки для параметров, передаваемых в функцию `customizeOrderModalTemplate`:
```
const testModalRenderingParams = (...args) => {
  if (Array.isArray(args)) {
      return true;
  }

  args[0] = everything[cardId].title;
  args[1] = everything[cardId].price;
  args[2] = everything[cardId].details;
  args[3] = everything[cardId].fullsized;
  args[4] = everything[cardId].thumbs;

  return args;
};
```
Но я затрудняюсь ответить, правильно ли я её написал и как мне её применить :confused:

eimwe replied:
"+" Исправлено. Добавил дефолтные значения для параметров на этапе дестрктуризации

kerim1505k commented on src/js/catalog.js:
```diff
@@ -1 +1,193 @@
+const closeOrderModal = () => {
+    const modalCloseButton = document.querySelector('.modal__close');
+
+    modalCloseButton.addEventListener('click', () => {
```
Проверка на modalCloseButton

eimwe replied:
"+" Исправлено

kerim1505k commented on src/js/catalog.js:
```diff
@@ -1 +1,193 @@
+
+const customizeOrderModalTemplate = (title, price, details, fullsized, thumbs) => {
+    const orderModalTemplate = document.getElementById('modal-order-template');
+    const orderModal = orderModalTemplate.content.firstElementChild.cloneNode(true);
```
Проверка на orderModalTemplate

eimwe replied:
"+" Исправлено
UPD: заметил, что линтер ругается каждый раз, когда я проставляю оператор `?`. Причем, при добавлении нового аналогичного оператора в другой строке, с прежней строки ошибка исчезает. исполнению скрипта это не мешает. Просто мне стало интересно, это ожидаемое поведение линтера или я как-то неправильно проставляю символ `?` Например, линтер подсвечивает ошибкой эту строку:
`cardTemplate?.content.firstElementChild.cloneNode(true);`

eimwe replied:
UPD 2021-01-12: проблема с линтером больше неактуальна, перезапуск IDE все вылечил

kerim1505k commented on src/js/catalog.js:
```diff
@@ -1 +1,193 @@
+const customizeOrderModalTemplate = (title, price, details, fullsized, thumbs) => {
+    const orderModalTemplate = document.getElementById('modal-order-template');
+    const orderModal = orderModalTemplate.content.firstElementChild.cloneNode(true);
+    const fullsizedImgWrapper = orderModal.querySelector('.gallery__fullsize');
```
Проверки на orderModal

eimwe replied:
"+" Исправлено

kerim1505k commented on src/js/catalog.js:
```diff
@@ -1 +1,193 @@
+    orderModal.querySelector('.modal__price').textContent = `${price} баллов`;
+    orderModal.querySelectorAll('.modal__details')[0].textContent = details;
+
+    fullsized.forEach((fullImage) => {
```
Очень похожий друг на друга код с 49 по 73 строку, сможешь объединить в функцию?

eimwe replied:
"+" Исправил. Объединил в функцию addGalleryImages с всеми параметрами, необходимыми для корректной вставки как тамбнейлов, так и полноразмерных фото

kerim1505k reviewed (changes requested):
Нужны проверки на элементы, которые берешь через querySelector и на передаваемые в функции параметры. Замоканные массивы статичны и в них можно быть уверенными, но в реальных проектах данные с бэкенда бывают разными, поэтому дополнительные проверки на их наличие, либо присвоение параметрам дефолтных значений является обязательным, чтобы работа скрипта не обвалилась.

eimwe committed:
Добавил изменения в соответствии с комментариями

1. Добавил оператор '?.' для проверки узлов, получаемых из DOM
2. Добавил правку в файл .eslintrc, чтобы избежать ошибки при проверке оператора '?.'
3. Добавил функцию addGalleryImages() для обработки массивов с картинками для модального окна
4. Добавил значения по умолчанию для параметров, передаваемых в функции генерации карточек товара и модальных окон
5. Избавился от излишнего кода в указанных строках
